File: yo-cordapp/java-source/src/test/java/net/corda/yo/YoContractTests.java
package net.corda.yo;

import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.CommandData;
import net.corda.core.identity.CordaX500Name;
import net.corda.testing.contracts.DummyContract;
import net.corda.testing.contracts.DummyState;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;
import org.junit.Test;


import static net.corda.testing.node.NodeTestUtils.ledger;

public class YoContractTests {
    MockServices ledgerServices = new MockServices(ImmutableList.of("net.corda.yo", "net.corda.testing.contracts"));
    TestIdentity alice = new TestIdentity(new CordaX500Name("Alice", "New York", "US"));
    TestIdentity bob = new TestIdentity(new CordaX500Name("Bob", "Tokyo", "JP"));
    TestIdentity miniCorp = new TestIdentity(new CordaX500Name("MiniCorp", "New York", "US"));


    @Test
    public void yoTransactionMustBeWellFormed() {
        YoState yo = new YoState(alice.getParty(), bob.getParty());
        ledger(ledgerServices, (ledger -> {
            //Too many inputs
            ledger.transaction(tx -> {
                tx.input(DummyContract.PROGRAM_ID, new DummyState());
                tx.command(alice.getPublicKey(), new YoContract.Commands.Send());
                tx.output(YoContract.ID, yo);
                tx.failsWith("There can be no inputs when Yo'ing other parties");
                return null;
            });
            //Wrong Command
            ledger.transaction(tx -> {
                tx.output(YoContract.ID, yo);
                tx.command(alice.getPublicKey(), new DummyCommand());
                tx.failsWith("");
                return null;
            });
            // Command signed by wrong key
            ledger.transaction(tx -> {
               tx.output(YoContract.ID,yo);
               tx.command(miniCorp.getPublicKey(), new YoContract.Commands.Send());
               tx.failsWith("The Yo! must be signed by the sender.");
               return null;
            });

            //Sending to yourself is not allowed
            ledger.transaction(tx -> {
                tx.output(YoContract.ID, new YoState(alice.getParty(), alice.getParty()));
                tx.command(miniCorp.getPublicKey(), new YoContract.Commands.Send());
                tx.failsWith("No sending Yo's to yourself!");
                return null;
            });

            //You can do a transaction
            ledger.transaction(tx -> {
                tx.output(YoContract.ID, yo);
                tx.command(alice.getPublicKey(), new YoContract.Commands.Send());
                tx.verifies();
                return null;
            });
            return null;
        }));
    }
};


class DummyCommand implements CommandData {}

File: yo-cordapp/contracts-kotlin/src/test/kotlin/net/corda/yo/contract/YoContractTests.kt
package net.corda.yo.test.contract

import net.corda.core.identity.CordaX500Name
import net.corda.testing.contracts.DummyContract
import net.corda.testing.contracts.DummyState
import net.corda.testing.core.DummyCommandData
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import net.corda.yo.contract.YoContract
import net.corda.yo.state.YoState
import org.junit.Test



class YoContractTests {
    private val ledgerServices = MockServices(listOf("net.corda.yo.contract", "net.corda.yo.flow", "net.corda.testing.contracts"))
    private val alice = TestIdentity(CordaX500Name("Alice", "New York", "US"))
    private val bob = TestIdentity(CordaX500Name("Bob", "Tokyo", "JP"))
    private val miniCorp = TestIdentity(CordaX500Name("MiniCorp", "New York", "US"))

    @Test
    fun yoTransactionMustBeWellFormed() {
        // A pre-made Yo to Bob.
        val yo = YoState(alice.party, bob.party)
        // Tests.
        ledgerServices.ledger {
            // Input state present.
            transaction {
                input(DummyContract.PROGRAM_ID, DummyState())
                command(alice.publicKey, YoContract.Send())
                output(YoContract.ID, yo)
                this.failsWith("There can be no inputs when Yo'ing other parties.")
            }
            // Wrong command.
            transaction {
                output(YoContract.ID, yo)
                command(alice.publicKey, DummyCommandData)
                this.failsWith("")
            }
            // Command signed by wrong key.
            transaction {
                output(YoContract.ID, yo)
                command(miniCorp.publicKey, YoContract.Send())
                this.failsWith("The Yo! must be signed by the sender.")
            }
            // Sending to yourself is not allowed.
            transaction {
                output(YoContract.ID, YoState(alice.party, alice.party))
                command(alice.publicKey, YoContract.Send())
                this.failsWith("No sending Yo's to yourself!")
            }
            transaction {
                output(YoContract.ID, yo)
                command(alice.publicKey, YoContract.Send())
                this.verifies()
            }
        }
    }
}